xds/internal/xdsclient/xds.go

```diff
@@ -360,6 +366,58 @@ func generateRDSUpdateFromRouteConfiguration(rc *v3routepb.RouteConfiguration, l
 	return RouteConfigUpdate{VirtualHosts: vhs}, nil
 }
 
+func generateRetryConfig(rp *v3routepb.RetryPolicy) (*RetryConfig, error) {
+	if !env.RetrySupport || rp == nil {
+		return nil, nil
+	}
+
+	cfg := &RetryConfig{RetryOn: make(map[codes.Code]bool)}
+	for _, s := range strings.Split(rp.GetRetryOn(), ",") {
+		switch strings.TrimSpace(strings.ToLower(s)) {
+		case "cancelled":
+			cfg.RetryOn[codes.Canceled] = true
+		case "deadline-exceeded":
+			cfg.RetryOn[codes.DeadlineExceeded] = true
+		case "internal":
+			cfg.RetryOn[codes.Internal] = true
+		case "resource-exhausted":
+			cfg.RetryOn[codes.ResourceExhausted] = true
+		case "unavailable":
+			cfg.RetryOn[codes.Unavailable] = true
+		}
+	}
+	if len(cfg.RetryOn) == 0 {
+		return &RetryConfig{}, nil
+	}
+	if rp.NumRetries == nil {
+		cfg.NumRetries = 1
+	} else {
+		cfg.NumRetries = rp.GetNumRetries().Value
+		if cfg.NumRetries < 1 {
+			return nil, fmt.Errorf("retry_policy.num_retries = %v; must be >= 1", cfg.NumRetries)
+		}
+	}
+
+	backoff := rp.GetRetryBackOff()
+	if backoff == nil {
+		cfg.RetryBackoff.BaseInterval = 25 * time.Millisecond
+	} else {
+		cfg.RetryBackoff.BaseInterval = backoff.GetBaseInterval().AsDuration()
+		if cfg.RetryBackoff.BaseInterval <= 0 {
+			return nil, fmt.Errorf("retry_policy.base_interval = %v; must be > 0", cfg.RetryBackoff.BaseInterval)
+		}
+	}
+	if max := backoff.GetMaxInterval(); max == nil {
```

Comment:
Couldn't backoff be nil, then this would panic?

Reply:
`GetMaxInterval` (and proto `Get`ters in general) handles `nil` receivers gracefully.

Reply:
Noted. The one I was thinking about was Menghans comment on GetX().Y (<- this would panic), vs GetX().GetY() which wouldn't, like what you have here.

Reply:
I've seen a _lot_ of bugs using Go protobufs due to similar things.  You have to be super careful with this.  Especially noticing there is no `IsSet` method, which makes testing for unset scalars ([which _is/will be_ a thing again in proto3](https://github.com/protocolbuffers/protobuf/blob/master/docs/implementing_proto3_presence.md)) in sub-messages much harder:

```go
if x.GetY().GetZ().GetIntegerA() == 0 {
	// seems like it might work, but you don't know whether IntegerA
	// is actually unset or set to zero explicitly
} 

if x.GetY().GetZ().IntegerA == nil { } // sorry, but this panics if Z is unset

if x.GetY().GetZ() == nil || x.GetY().GetZ().IntegerA == nil {
	// now we know IntegerA is not set
}
```